favor/favor.py

```diff
@@ -87,22 +97,39 @@ def phi(x):
         # (batch, n, r)
         phi_q = phi(queries)
 
-        out = torch.matmul( # (batch, n, dim)
-            phi_q,
-            torch.matmul( # (batch, r, dim)
-                phi_k.permute(0, 2, 1), values
+        if self.causal:
+            # outer products of keys and values: (batch, n, r, dim)
+            k_v_prod = torch.matmul(phi_k[:, :, :, None], values[:, :, None, :])
+
+            out = torch.matmul(         # (batch, n, dim)
+                phi_q[:, :, None, :],   # (batch, n, 1, r)
+                k_v_prod.cumsum(dim=1)  # (batch, n, r, dim)
+            ).squeeze(2)
+
+            # normalization factors: (batch, n, 1)
+            norm = torch.matmul(
+                phi_q[:, :, None, :],           # (batch, n, 1, r)
+                phi_k.cumsum(dim=1)[..., None]  # (batch, n, r, 1)
+            ).squeeze(2)
```

Comment:
are you 100% sure ?

Reply:
Hmm, probably not 100%. :smile: We could write a test to check that we can approximate standard attention with a causal attention mask (with the softmax kernel).

Reply:
ok, what about memory usage ? we never store the big attention, ok ?

Reply:
Yes, that's OK, the `n` dimension is always one of the non-matrix ("batch") dimensions in the `matmul`. It's the trick from Appendix B.1.

Reply:
all right, awesome